feg/gateway/diameter/connection.go

```diff
@@ -155,16 +159,50 @@ func (c *Connection) getDiamConnection() (diam.Conn, *smpeer.Metadata, error) {
 		return nil, nil, errors.New("Could not obtain metadata from connection")
 	}
 	c.conn, c.metadata = conn, metadata
+	if cn, ok := conn.(diam.CloseNotifier); ok && cn != nil {
+		go c.connCloseNotify(cn.CloseNotify(), conn)
+	} else {
+		glog.Errorf("new diam conn (%T) %s -> %s is not CloseNotifier", conn, localAddr, c.server.Addr)
+	}
 	return conn, metadata, nil
 }
 
+func (c *Connection) connCloseNotify(cnc <-chan struct{}, conn diam.Conn) {
+	<-cnc // wait for close notifier
+	if glog.V(2) {
+		remoteAddress, localAddress := "nil", "nil"
+		if conn != nil {
+			if conn.LocalAddr() != nil {
+				localAddress = conn.LocalAddr().String()
+			}
+			if conn.LocalAddr() != nil {
+				remoteAddress = conn.RemoteAddr().String()
+			}
+		}
+		glog.Infof("notified of %s->%s connection closure", localAddress, remoteAddress)
+	}
+	if c != nil && conn != nil && c.destroyConnection(conn) {
+		// if connection was closed not by connection management functions, recover it
+		retryWaitTime := connectionRecoveryInterval
+		for i := 0; i < connectionRecoveryattempts; i++ {
+			_, _, err := c.getDiamConnection()
+			if err == nil {
+				break
+			}
+			time.Sleep(retryWaitTime)
+			retryWaitTime *= 2
```

Comment:
now i am just concerned about HA functionality :D

Reply:
HE disable will set connection to disabled...